numba/_typeof.c

```diff
@@ -0,0 +1,730 @@
+#include "_pymodule.h"
+
+#include <string.h>
+#include <time.h>
+#include <assert.h>
+
+#include "_typeof.h"
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include <numpy/ndarrayobject.h>
+
+
+/* Cached typecodes for basic scalar types */
+static int tc_int8;
+static int tc_int16;
+static int tc_int32;
+static int tc_int64;
+static int tc_uint8;
+static int tc_uint16;
+static int tc_uint32;
+static int tc_uint64;
+static int tc_float32;
+static int tc_float64;
+static int tc_complex64;
+static int tc_complex128;
+static int BASIC_TYPECODES[12];
+
+static int tc_intp;
+
+static PyObject *typecache;
+static PyObject *ndarray_typecache;
+static PyObject *structured_dtypes;
+
+static PyObject *str_typeof_pyval = NULL;
+
+
+/*
+ * Type fingerprint computation.
+ */
+
+typedef struct {
+    char *buf;
+    size_t n;
+    size_t allocated;
+    char static_buf[40];
```

Comment:
> why 40 char?

So that the structure fits in a cache line. It's mostly pedantic, though. The important thing is that most usual types will fit in the static buf.